source/hud.lua

```diff
@@ -141,18 +141,22 @@ local function drawShopMenu()
 		Assets.setFont("font20")
 
 		-- Create List of available modules
-		for _, module in pairs(Modules) do
-			local string = "%s. Buy %s - $%s \n"
-			itemListString = string.format(string, module.id, module.name, module.cost)
-			-- Draw list of modules
-			local color = {1, 1, 1, 1}
-			local y = SCREEN_HEIGHT * 0.33
-			if Lander.hasUpgrade(LANDERS[1], module) then
-				color = {.8, .1, .1, .5}
+		local y = SCREEN_HEIGHT * 0.33
+
+		for k, module in pairs(Modules) do
+			if module.allowed == nil or module.allowed == true then
+				local string = "%s. Buy %s - $%s \n"
+				itemListString = string.format(string, module.id, module.name, module.cost)
+				-- Draw list of modules
+				local color = {1, 1, 1, 1}
+				if Lander.hasUpgrade(LANDERS[1], module) then
+					color = {.8, .1, .1, .5}
+				end
+				love.graphics.setColor(color)
+				love.graphics.printf(itemListString, 0, y, SCREEN_WIDTH, "center")
+				y = y + 20
```

Comment:
As already discussed, we don't need to sort the table to show the items in the correct order. You just removed the code I already added to set the drawing order by the item id:
```lua
love.graphics.printf(itemListString, 0, y + (20*module.id), SCREEN_WIDTH, "center")
```

I acknowledge that there is an issue if we remove an item (causes a gap where a entry is missing), but as I stated, I don't think we should remove anything anyway - and if we really need to, we should make sure that we iterate another table with modules instead of the source table `Modules` directly. One solution might be to use the lander.modules table since this one is already numerical indexed and we don't use it to point to a specific module.
For that cause, we also should get rid of the connected index for shop entry and module.id (which is used for `keypressed` to buy it), because this is just messy and makes things more difficult.